### PR TITLE
Pallet liquidity rewards impl

### LINKED ISSUE #999: Rewards: pallet-liquidity-rewards implementation
### Which part of the code is the issue addressing?

* Feature: https://github.com/centrifuge/roadmap/issues/163

### Description

Implements the `pallet-liquidity-rewards`. This implies:
- Implementation
- Testing
- Benchmarks

### Research/based on
- See [specs](https://centrifuge.hackmd.io/9T9zAZaTRD6ZcHN-Gd_lzw?view)

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

Implements [`liquidity-pallet-rewards`](https://centrifuge.hackmd.io/9T9zAZaTRD6ZcHN-Gd_lzw?view#pallet-liquidity-rewards-spec)

Fixes #999

## Changes and Descriptions

- [x] Implement extrinsics
- [x] on_initialized hook
- [x] Weights
- [x] Mock
- [x] Tests
- [x] Events
- [x] Docs
- [x] Readme doc entry
- [x] Benchmarks

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

```sh
cargo test -p pallet-liquidity-rewards --features runtime-benchmarks
PARA_CHAIN_SPEC=development ./scripts/init.sh benchmark pallet-liquidity-rewards
```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I rebased on the latest `main` branch

### CODE REVIEW

**Review by mikiquantum — COMMENTED** [review]
Overall is good and simple, I will have another look as soon as the other tasks listed in the description are in 👍

File: pallets/liquidity-rewards/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		#[pallet::constant]
+		type MaxChangesPerEpoch: Get<Option<u32>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(super) type GroupWeights<T: Config> =
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/liquidity-rewards/src/lib.rs:111]: The `Groups` storage of `pallet-rewards` stores the `Group` data for each `GroupId` (without weight information).

This new storage does two things:
  - Stores the weight of each group
  - Identify the `GroupId` used only by this pallet. The length of this storage can be (and will be) different to the length of `Groups` in `pallet-rewards` where all groups used in other pallets will be stored there.

File: pallets/liquidity-rewards/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+				Ok(())
+			})
+			.ok();
+
+			func_weight
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/liquidity-rewards/src/lib.rs:200]: Totally, the only "complicated" place is `on_initialize()`.

File: pallets/liquidity-rewards/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikiquantum) [reply pallets/liquidity-rewards/src/lib.rs:111]: ok, I see, I saw the T::Rate type in the pallet rewards, and thought that was the weight, but is the rpt. It makes sense that the rewards pallet doesnt need to know about an implementation that uses weights for distribution.

**Comment by lemunozm** [comment]
@Miguel changes from your last review:
- Added a mock for rewards traits using [`mockall`](https://docs.rs/mockall/latest/mockall/index.html) (a cool mocking library).
- Simplified `mock.rs` a lot (thanks to mocking the reward trait).
- Added tests/benchmarks/docs.
- Modified some storage to improve the performance at `on_initialize()`
  - `GroupWeights` storage is now a simple `BoundedBTreeMap` in a `StorageValue`.
  - `EpochChanges` is a new struct that combines all different epoch changes and is based on `BoundedBTreeMap` instead of using several `StorageMap`.
  - `ends_on` attribute of epoch extract to its own struct to reduce the deserialization phase each `on_initialize`. There is no need to deserialize the whole epoch if 99% of `on_initialize()` calls will not use the rest of the `Epoch` data.
- Added `pallet-liquidity-pallet` to `development` runtime (needed to generate weights)

File: pallets/liquidity-rewards/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+		///
+		/// This method will do the currency available for using it in stake/unstake/claim calls.
+		#[pallet::weight(T::WeightInfo::set_currency_group())]
+		pub fn set_currency_group(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			group_id: T::GroupId,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			NextEpochChanges::<T>::try_mutate(|changes| {
```
Comment (mikiquantum) [inline pallets/liquidity-rewards/src/lib.rs:402]: nit: spacing

File: pallets/liquidity-rewards/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+		/// Information of runtime weights
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(super) type EndOfEpoch<T: Config> = StorageValue<
```
Comment (mikiquantum) [inline pallets/liquidity-rewards/src/lib.rs:200]: Maybe we can add a comment explaining that we took the endOfEpoch out of the active epoch data storage to not have to deserialize the whole struct for 99% of the time assuming a duration of 100 blocks :)

**Review by mikiquantum — APPROVED** [review]
LGTM!

File: pallets/liquidity-rewards/src/lib.rs (hunk 4 above)
Comment (lemunozm) [reply pallets/liquidity-rewards/src/lib.rs:200]: I like it!

File: runtime/development/Cargo.toml [hunk 5]
```diff
@@ -287,6 +292,7 @@ try-runtime = [
     "pallet-fees/try-runtime",
     "pallet-identity/try-runtime",
     "pallet-interest-accrual/try-runtime",
+    "pallet-liquidity-rewards/runtime-benchmarks",
```
Comment (mustermeiszer) [inline runtime/development/Cargo.toml:295]: We must be a bit more cautious with feature flags @lemunozm and @mikiquantum. I guess development would not build with try-runtime in this scenario separately. Just am marking this here, as these feature flags cost us quite some time, then they are set wrongly and some way down the line in the future make problems. 

This also goes for myself. Maybe we should brainstorm, how we could enforce correct feature flags and what should be the default setup for pallets and their features.

**Pull request merged**